### PR TITLE
[Merged by Bors] - chore: adaptations for nightly-2024-02-29

### PR DESCRIPTION
**Pull request by kim-em**
`field_simp` is a very ambitious tactic, that tries to prove that it can clear denominators. Previously it was taking advantage of essentially a bug in the simp discharging framework, so that parts of its discharging strategy were not limited by `maxDischargeDepth` (so it could "keep trying"), but everyone else was limited (so `simp` didn't waste time exploring large trees of discharges). Now the `maxDischargeDepth` is enforced for everyone, so to keep `field_simp` working we have to increase its `maxDischargeDepth`. That makes it extremely slow. I've kept the `maxDischargeDepth` to a minimum, but still there are some `set_option maxHeartbeats` (all with notes attached).

After we've merged this, someone is going to have to rewrite `field_simp` to make it performant.

-------

#11079 is backporting a few changes to master

### CODE REVIEW

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 1]
```diff
@@ -693,7 +693,7 @@ instance [DecidableEq E] : One (NonarchAddGroupSeminorm E) :=
         · simp_rw [if_pos hx, hx, zero_add]
           exact le_max_of_le_right (le_refl _)
         · simp_rw [if_neg hx]
-          split_ifs <;> norm_num
+          split_ifs <;> simp
```
Comment (grunweg) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:696]: What happened here? (It's not obvious to me - but I'm not an expert; feel free to discard this comment.)

File: Mathlib/CategoryTheory/Limits/Shapes/ZeroMorphisms.lean [hunk 2]
```diff
@@ -220,7 +220,7 @@ theorem iff_isSplitMono_eq_zero {X Y : C} (f : X ⟶ Y) [IsSplitMono f] : IsZero
     rw [← Category.id_comp f, h, zero_comp]
   · intro h
     rw [← IsSplitMono.id f]
-    simp [h]
+    simp only [h, zero_comp]
```
Comment (grunweg) [inline Mathlib/CategoryTheory/Limits/Shapes/ZeroMorphisms.lean:223]: I think this is a good change, just asking for curiosity: (why) is this needed? Again, feel free to discard.

File: Mathlib/Algebra/ContinuedFractions/Computation/CorrectnessTerminating.lean [hunk 3]
```diff
@@ -132,8 +132,8 @@ theorem compExactValue_correctness_of_stream_eq_some :
     -- Nat.succ
     obtain ⟨ifp_n, nth_stream_eq, nth_fract_ne_zero, -⟩ :
       ∃ ifp_n, IntFractPair.stream v n = some ifp_n ∧
-        ifp_n.fr ≠ 0 ∧ IntFractPair.of ifp_n.fr⁻¹ = ifp_succ_n :=
-      IntFractPair.succ_nth_stream_eq_some_iff.1 succ_nth_stream_eq
+        ifp_n.fr ≠ 0 ∧ IntFractPair.of ifp_n.fr⁻¹ = ifp_succ_n
+    exact IntFractPair.succ_nth_stream_eq_some_iff.1 succ_nth_stream_eq
```
Comment (jcommelin) [inline Mathlib/Algebra/ContinuedFractions/Computation/CorrectnessTerminating.lean:136]: This seems to be reverting to flow-of-consciousness `obtain` proofs.
(Also in the rest of this file.)

File: Mathlib/Algebra/Free.lean [hunk 4]
```diff
@@ -536,6 +536,7 @@ def lift : (α → β) ≃ (FreeSemigroup α →ₙ* β) where
       map_mul' := fun x y ↦ by
         simp [head_mul, tail_mul, ← List.foldl_map, List.foldl_append, List.foldl_cons,
           List.foldl_assoc] }
+        -- v4.7.0 issues
```
Comment (jcommelin) [inline Mathlib/Algebra/Free.lean:539]: This is the only change in this file... can we just drop it?
```suggestion
```

File: Mathlib/LinearAlgebra/FreeModule/Norm.lean [hunk 5]
```diff
@@ -29,7 +29,7 @@ free as an `R`-module, the norm of `f` relative to `R` is associated to the prod
 coefficients of the ideal generated by `f`. -/
 theorem associated_norm_prod_smith [Fintype ι] (b : Basis ι R S) {f : S} (hf : f ≠ 0) :
     Associated (Algebra.norm R f) (∏ i, smithCoeffs b _ (span_singleton_eq_bot.not.2 hf) i) := by
-  have hI := span_singleton_eq_bot.not.2 hf
+  have hI := span_singleton_eq_bot.not.2 hf -- v4.7.0-rc1 issues
```
Comment (grunweg) [inline Mathlib/LinearAlgebra/FreeModule/Norm.lean:32]: Is this one still needed?

File: Mathlib/Order/Filter/AtTopBot.lean [hunk 6]
```diff
@@ -1907,7 +1907,7 @@ theorem exists_seq_monotone_tendsto_atTop_atTop (α : Type*) [SemilatticeSup α]
   obtain ⟨ys, h⟩ := exists_seq_tendsto (atTop : Filter α)
   let xs : ℕ → α := fun n => Finset.sup' (Finset.range (n + 1)) Finset.nonempty_range_succ ys
   have h_mono : Monotone xs := fun i j hij ↦ by
-    simp only [xs] -- Need to unfold `xs` and do alpha reduction, otherwise `gcongr` fails
+    simp only [xs]
```
Comment (grunweg) [inline Mathlib/Order/Filter/AtTopBot.lean:1910]: Is this change intentional?

File: Mathlib/Analysis/SpecificLimits/Normed.lean [hunk 7]
```diff
@@ -716,7 +716,7 @@ theorem Monotone.tendsto_le_alternating_series
       ← sub_eq_add_neg, sub_le_iff_le_add]
     gcongr
     exact hfm (by omega)
-  exact ha.le_of_tendsto (hfl.comp (tendsto_atTop_mono (fun n ↦ by omega) tendsto_id)) _
+  exact ha.le_of_tendsto (hfl.comp (tendsto_atTop_mono (fun n ↦ by dsimp; omega) tendsto_id)) _
```
Comment (jcommelin) [inline Mathlib/Analysis/SpecificLimits/Normed.lean:719]: I'm happy with this fix, just flagging that it isn't the `set_option` variant that we see almost everywhere else.

File: Mathlib/RingTheory/Polynomial/ScaleRoots.lean [hunk 8]
```diff
@@ -226,6 +227,12 @@ lemma mul_scaleRoots_of_noZeroDivisors (p q : R[X]) (r : R) [NoZeroDivisors R] :
 lemma add_scaleRoots_of_natDegree_eq (p q : R[X]) (r : R) (h : natDegree p = natDegree q) :
     r ^ (natDegree p - natDegree (p + q)) • (p + q).scaleRoots r =
       p.scaleRoots r + q.scaleRoots r := by
+  -- Was:
```
Comment (grunweg) [inline Mathlib/RingTheory/Polynomial/ScaleRoots.lean:230]: Is this one intentional?

File: Mathlib/Data/Nat/Digits.lean [hunk 9]
```diff
@@ -729,7 +729,7 @@ theorem eleven_dvd_iff :
 
 theorem eleven_dvd_of_palindrome (p : (digits 10 n).Palindrome) (h : Even (digits 10 n).length) :
     11 ∣ n := by
-  let dig := (digits 10 n).map fun n : ℕ => (n : ℤ)
+  let dig := (digits 10 n).map (fun i : ℕ => (i : ℤ))
```
Comment (jcommelin) [inline Mathlib/Data/Nat/Digits.lean:732]: (why) do we need this change?

File: Mathlib/Data/Real/EReal.lean [hunk 10]
```diff
@@ -203,10 +203,11 @@ theorem induction₂ {P : EReal → EReal → Prop} (top_top : P ⊤ ⊤) (top_p
 /-- Induct on two `EReal`s by performing case splits on the sign of one whenever the other is
 infinite. This version eliminates some cases by assuming that the relation is symmetric. -/
 @[elab_as_elim]
-theorem induction₂_symm {P : EReal → EReal → Prop} (symm : Symmetric P) (top_top : P ⊤ ⊤)
-    (top_pos : ∀ x : ℝ, 0 < x → P ⊤ x) (top_zero : P ⊤ 0) (top_neg : ∀ x : ℝ, x < 0 → P ⊤ x)
-    (top_bot : P ⊤ ⊥) (pos_bot : ∀ x : ℝ, 0 < x → P x ⊥) (coe_coe : ∀ x y : ℝ, P x y)
-    (zero_bot : P 0 ⊥) (neg_bot : ∀ x : ℝ, x < 0 → P x ⊥) (bot_bot : P ⊥ ⊥) : ∀ x y, P x y :=
+theorem induction₂_symm {P : EReal → EReal → Prop} (symm : ∀ {x y}, P x y → P y x)
```
Comment (jcommelin) [inline Mathlib/Data/Real/EReal.lean:206]: Does `Symmetric` no longer exist? Or why are we making this change?

**Review by grunweg — COMMENTED** [review]
A few non-expert comments - it seems this could use a merge of mathlib's master branch.

I was wondering: are the remaining `Adaptation note`s supposed to stay? If so, would a tracking issue for them be useful, so they don't get forgotten?

File: Mathlib/LinearAlgebra/FreeModule/Norm.lean (hunk 5 above)
Comment (jcommelin) [inline Mathlib/LinearAlgebra/FreeModule/Norm.lean:32]: Can we drop this comment?

File: Mathlib/MeasureTheory/Integral/SetToL1.lean [hunk 11]
```diff
@@ -1157,8 +1157,7 @@ theorem setToL1_mono_left' {T T' : Set α → E →L[ℝ] G''} {C C' : ℝ}
     (hT : DominatedFinMeasAdditive μ T C) (hT' : DominatedFinMeasAdditive μ T' C')
     (hTT' : ∀ s, MeasurableSet s → μ s < ∞ → ∀ x, T s x ≤ T' s x) (f : α →₁[μ] E) :
     setToL1 hT f ≤ setToL1 hT' f := by
-  induction f using Lp.induction with
-  | hp_ne_top h => exact one_ne_top h
+  induction f using Lp.induction (hp_ne_top := one_ne_top) with
```
Comment (jcommelin) [inline Mathlib/MeasureTheory/Integral/SetToL1.lean:1160]: I'm fine with this change. But why is it part of this PR?

File: Mathlib/Order/BooleanAlgebra.lean [hunk 12]
```diff
@@ -852,6 +852,7 @@ protected def Function.Injective.generalizedBooleanAlgebra [Sup α] [Inf α] [Bo
     (map_sup : ∀ a b, f (a ⊔ b) = f a ⊔ f b) (map_inf : ∀ a b, f (a ⊓ b) = f a ⊓ f b)
     (map_bot : f ⊥ = ⊥) (map_sdiff : ∀ a b, f (a \ b) = f a \ f b) :
     GeneralizedBooleanAlgebra α where
+  __ := hf.distribLattice f map_sup map_inf
```
Comment (jcommelin) [inline Mathlib/Order/BooleanAlgebra.lean:855]: Why do we need this now?

File: Mathlib/Order/CompactlyGenerated/Basic.lean [hunk 13]
```diff
@@ -465,6 +465,7 @@ lemma CompleteLattice.independent_iff_supIndep_of_injOn {ι : Type*} {f : ι →
     refine ⟨by aesop, fun ⟨ha, has⟩ ↦ ?_⟩
     obtain ⟨j, hj, rfl⟩ := hs has
     exact ⟨j, ⟨hj, ha, has⟩, rfl⟩
+  change Disjoint (f i) (Finset.sup t id)
```
Comment (jcommelin) [inline Mathlib/Order/CompactlyGenerated/Basic.lean:468]: Why do we need this now?

File: Mathlib/Order/Filter/AtTopBot.lean (hunk 6 above)
Comment (jcommelin) [inline Mathlib/Order/Filter/AtTopBot.lean:1910]: Why are we dropping the comment?

File: Mathlib/RingTheory/Polynomial/ScaleRoots.lean [hunk 14]
```diff
@@ -226,6 +227,12 @@ lemma mul_scaleRoots_of_noZeroDivisors (p q : R[X]) (r : R) [NoZeroDivisors R] :
...
+  -- ext n; simp only [coeff_smul, coeff_scaleRoots, coeff_add, smul_eq_mul,
+  --   mul_comm (r ^ _), mul_assoc, ← pow_add, ← h, ← add_mul, add_comm (_ - n)]
+  -- cases lt_or_le (natDegree (p + q)) n with
+  -- | inl hn => simp only [← coeff_add, coeff_eq_zero_of_natDegree_lt hn, zero_mul]
+  -- | inr hn => rw [tsub_add_tsub_cancel (natDegree_add_le_of_degree_le le_rfl h.ge) hn]
```
Comment (jcommelin) [inline Mathlib/RingTheory/Polynomial/ScaleRoots.lean:235]: Can we drop this? It doesn't seem to correspond to code in master...

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/EulerSineProd.lean [hunk 15]
```diff
@@ -158,7 +158,8 @@ theorem integral_cos_mul_cos_pow (hn : 2 ≤ n) (hz : z ≠ 0) :
   have := integral_cos_mul_cos_pow_aux hn hz
   rw [integral_sin_mul_sin_mul_cos_pow_eq hn hz, sub_eq_neg_add, mul_add, ← sub_eq_iff_eq_add]
     at this
-  convert congr_arg (fun u : ℂ => -u * (2 * z) ^ 2 / n ^ 2) this using 1 <;> field_simp <;> ring
+  convert congr_arg (fun u : ℂ => -u * (2 * z) ^ 2 / n ^ 2) this using 1
+    <;> field_simp <;> ring
```
Comment (jcommelin) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/EulerSineProd.lean:162]: I guess this line break can be reverted again?
```suggestion
  convert congr_arg (fun u : ℂ => -u * (2 * z) ^ 2 / n ^ 2) this using 1 <;> field_simp <;> ring
```

File: time_startup.sh [hunk 16]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (jcommelin) [inline time_startup.sh:1]: Do you intend to keep this script? If so, please move it to `scripts/` and add 3 lines of docs explaining what it is for.

File: Mathlib/Algebra/ContinuedFractions/Computation/CorrectnessTerminating.lean [hunk 17]
```diff
@@ -132,8 +132,8 @@ theorem compExactValue_correctness_of_stream_eq_some :
     -- Nat.succ
     obtain ⟨ifp_n, nth_stream_eq, nth_fract_ne_zero, -⟩ :
       ∃ ifp_n, IntFractPair.stream v n = some ifp_n ∧
-        ifp_n.fr ≠ 0 ∧ IntFractPair.of ifp_n.fr⁻¹ = ifp_succ_n :=
-      IntFractPair.succ_nth_stream_eq_some_iff.1 succ_nth_stream_eq
+        ifp_n.fr ≠ 0 ∧ IntFractPair.of ifp_n.fr⁻¹ = ifp_succ_n
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Algebra/ContinuedFractions/Computation/CorrectnessTerminating.lean:135]: Yes: the process here is merging `master` to `bump/v4.7.0`, and then merging `bump/v4.7.0` into `bump/nightly-202402-29`. I've just done that, so hopefully these diffs disappear.

File: time_startup.sh (hunk 16 above)
Comment (semorrison) [reply time_startup.sh:1]: No, this was a stray.

File: Archive/Examples/IfNormalization/WithoutAesop.lean [hunk 18]
```diff
@@ -99,8 +99,8 @@ def normalize' (l : AList (fun _ : ℕ => Bool)) :
           · subst h'
             simp_all
           · simp_all? says simp_all only [hasNestedIf, Bool.or_self, hasConstantIf, and_self,
-              hasRedundantIf, Bool.or_false, beq_eq_false_iff_ne, ne_eq, not_false_eq_true,
-              disjoint, List.disjoint, decide_True, Bool.and_self]
+                hasRedundantIf, Bool.or_false, beq_eq_false_iff_ne, ne_eq, not_false_eq_true,
```
Comment (grunweg) [inline Archive/Examples/IfNormalization/WithoutAesop.lean:102]: Why this indentation change?

File: Archive/Wiedijk100Theorems/SolutionOfCubic.lean [hunk 19]
```diff
@@ -101,16 +101,18 @@ theorem cubic_eq_zero_iff (ha : a ≠ 0) (hω : IsPrimitiveRoot ω 3)
...
-  have hi2 : (2 : K) ≠ 0 := nonzero_of_invertible _
   have hi3 : (3 : K) ≠ 0 := nonzero_of_invertible _
   have h9 : (9 : K) = 3 ^ 2 := by norm_num
   have h54 : (54 : K) = 2 * 3 ^ 3 := by norm_num
   have h₁ : a * x ^ 3 + b * x ^ 2 + c * x + d
     = a * (x ^ 3 + b / a * x ^ 2 + c / a * x + d / a) := by field_simp; ring
   have h₂ : ∀ x, a * x = 0 ↔ x = 0 := by intro x; simp [ha]
-  have hp' : p = (3 * (c / a) - (b / a) ^ 2) / 9 := by field_simp [hp, h9]; ring_nf
+  have hp' : p = (3 * (c / a) - (b / a) ^ 2) / 9 := by
+    field_simp [hp, h9]; ring_nf
```
Comment (grunweg) [inline Archive/Wiedijk100Theorems/SolutionOfCubic.lean:111]: Is there reason for this line break? Can you revert this?

File: Archive/Wiedijk100Theorems/SolutionOfCubic.lean [hunk 20]
```diff
@@ -140,7 +142,8 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
         field_simp; ring
       _ = a * (x + b / (3 * a)) ^ 3 + (d - (9 * a * b * c - 2 * b ^ 3) * a / (3 * a) ^ 3) := by
         simp only [hb2, hb3]; field_simp; ring
-      _ = a * ((x + b / (3 * a)) ^ 3 - s ^ 3) := by rw [hs3, hq]; field_simp [h54]; ring
+      _ = a * ((x + b / (3 * a)) ^ 3 - s ^ 3) := by
+        rw [hs3, hq]; field_simp [h54]; ring
```
Comment (grunweg) [inline Archive/Wiedijk100Theorems/SolutionOfCubic.lean:146]: Idem: is this line break needed?

File: Mathlib/RingTheory/AlgebraTower.lean [hunk 21]
```diff
@@ -154,7 +154,9 @@ noncomputable def Basis.smul {ι : Type v₁} {ι' : Type w₁} (b : Basis ι R
 
 @[simp]
 theorem Basis.smul_repr {ι : Type v₁} {ι' : Type w₁} (b : Basis ι R S) (c : Basis ι' S A) (x ij) :
-    (b.smul c).repr x ij = b.repr (c.repr x ij.2) ij.1 := by simp [Basis.smul]; rfl
+    (b.smul c).repr x ij = b.repr (c.repr x ij.2) ij.1 := by
+  set_option tactic.skipAssignedInstances false in
```
Comment (grunweg) [inline Mathlib/RingTheory/AlgebraTower.lean:158]: Can you indent these lines again?

**Review by grunweg — COMMENTED** [review]
I've gone over the entire diff: apart from these few nits and the ones already mentioned, the changes are either mechanical, improvements or reverting temporary sorries.

File: Mathlib/MeasureTheory/Integral/SetToL1.lean (hunk 11 above)
Comment (grunweg) [reply Mathlib/MeasureTheory/Integral/SetToL1.lean:1160]: I'm backporting this in #11079.

File: Mathlib/CategoryTheory/Limits/Shapes/ZeroMorphisms.lean (hunk 2 above)
Comment (grunweg) [reply Mathlib/CategoryTheory/Limits/Shapes/ZeroMorphisms.lean:223]: I'm backporting this in #11079.

File: Mathlib/Algebra/ContinuedFractions/Computation/CorrectnessTerminating.lean [hunk 22]
```diff
@@ -143,7 +143,7 @@ theorem compExactValue_correctness_of_stream_eq_some :
     · suffices v = conts.a / conts.b by simpa [compExactValue, ifp_succ_n_fr_eq_zero]
       -- use the IH and the fact that ifp_n.fr⁻¹ = ⌊ifp_n.fr⁻¹⌋ to prove this case
       obtain ⟨ifp_n', nth_stream_eq', ifp_n_fract_inv_eq_floor⟩ :
-          ∃ ifp_n, IntFractPair.stream v n = some ifp_n ∧ ifp_n.fr⁻¹ = ⌊ifp_n.fr⁻¹⌋ :=
+        ∃ ifp_n, IntFractPair.stream v n = some ifp_n ∧ ifp_n.fr⁻¹ = ⌊ifp_n.fr⁻¹⌋ :=
```
Comment (grunweg) [inline Mathlib/Algebra/ContinuedFractions/Computation/CorrectnessTerminating.lean:146]:
```suggestion
          ∃ ifp_n, IntFractPair.stream v n = some ifp_n ∧ ifp_n.fr⁻¹ = ⌊ifp_n.fr⁻¹⌋ :=

File: Mathlib/Order/BooleanAlgebra.lean (hunk 12 above)
Comment (grunweg) [reply Mathlib/Order/BooleanAlgebra.lean]: We don't: removed.

File: Mathlib/Analysis/SpecificLimits/Normed.lean (hunk 7 above)
Comment (semorrison) [reply Mathlib/Analysis/SpecificLimits/Normed.lean:719]: We decide that `omega` can't afford to do identification of atoms up to defeq: in some problems it's too expensive.

So slightly more preprocessing will be needed in some cases.

File: Mathlib/Data/Nat/Digits.lean (hunk 9 above)
Comment (semorrison) [reply Mathlib/Data/Nat/Digits.lean]: Apparently not, reverted. (And cherry-picked back to nightly-testing)

**Comment by jcommelin** [comment]
Thanks :tada: 

If CI passes, please remove the label `awaiting-CI` and merge this yourself, by adding a comment `bors r+`.

bors d+

**Comment by semorrison** [comment]
bors p=10

**Comment by adomani** [comment]
I just cherry-picked [51da0513b468f47b974b0aa874ca730371f70416](https://github.com/leanprover-community/mathlib4/pull/11107/commits/51da0513b468f47b974b0aa874ca730371f70416), since it seems to fix `compute_degree` and not cause further breakage.

**Pull request closed without merging**